internal/types/types.go

```diff
@@ -0,0 +1,912 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/ydb-platform/ydb-go-genproto/protos/Ydb"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/allocator"
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xstring"
+)
+
+type Type interface {
+	Yql() string
+	String() string
+
+	ToYDB(a *allocator.Allocator) *Ydb.Type
+	equalsTo(rhs Type) bool
+}
+
+func TypeToYDB(t Type, a *allocator.Allocator) *Ydb.Type {
+	return t.ToYDB(a)
+}
+
+func TypeFromYDB(x *Ydb.Type) Type {
+	switch v := x.Type.(type) {
+	case *Ydb.Type_TypeId:
+		return primitiveTypeFromYDB(v.TypeId)
+
+	case *Ydb.Type_OptionalType:
+		return NewOptional(TypeFromYDB(v.OptionalType.Item))
+
+	case *Ydb.Type_ListType:
+		return NewList(TypeFromYDB(v.ListType.Item))
+
+	case *Ydb.Type_DecimalType:
+		d := v.DecimalType
+
+		return NewDecimal(d.Precision, d.Scale)
+
+	case *Ydb.Type_TupleType:
+		t := v.TupleType
+
+		return NewTuple(FromYDB(t.Elements)...)
+
+	case *Ydb.Type_StructType:
+		s := v.StructType
+
+		return NewStruct(StructFields(s.Members)...)
+
+	case *Ydb.Type_DictType:
+		keyType, valueType := TypeFromYDB(v.DictType.Key), TypeFromYDB(v.DictType.Payload)
+		if valueType.equalsTo(NewVoid()) {
+			return NewSet(keyType)
+		}
+
+		return NewDict(keyType, valueType)
+
+	case *Ydb.Type_VariantType:
+		t := v.VariantType
+		switch x := t.Type.(type) {
+		case *Ydb.VariantType_TupleItems:
+			return NewVariantTuple(FromYDB(x.TupleItems.Elements)...)
+		case *Ydb.VariantType_StructItems:
+			return NewVariantStruct(StructFields(x.StructItems.Members)...)
+		default:
+			panic("ydb: unknown variant type")
+		}
+
+	case *Ydb.Type_VoidType:
+		return NewVoid()
+
+	case *Ydb.Type_NullType:
+		return NewNull()
+
+	default:
+		panic("ydb: unknown type")
```

Comment:
Do you really need panic instead of error for unknown type when parse protobuf?

Reply:
This is legacy =(
The best way need a breaking change of function signature - one function result change to two results (value + error)